### ISSUE TITLE
[dv] Timescale in DV

### ISSUE THREAD

**Issue by weicaiyang** [lowRISC/opentitan#11319]
In most of the TBs, we use 1ns/1ps timescale for DV components. But @Jacob-Levy and I found a test in private repo, where the DV timescale inherited a 1ps/1ps timescale from a standard cell.

Not sure what the best way to deal with the timescale in DV.
1. make DV timescale agnostic
We do try to make our DV code timescale agnostic, but we don't really test the code with multiple timescale setting. So, we may see this kind of issue (#11318) again when the DV timescale isn't 1ns/1ps.
But with that fix, we can have the chip uart test pass now. Maybe it's not too bad to go this route.

2. use a fixed timescale and avoid being overridden. 
Don't really know if there is an easy way to achieve this. Or, we may need to add timescale in every DV package.

Any thoughts?

@sriyerg @matutem @cindychip @rswarbrick @rasmus-madsen @Jacob-Levy

**Comment by moidx** [comment]
Keeping as part of M2.5 to follow up with @sha-ron.

**Comment by johngt** [comment]
Reassigning based on previous description.

**Comment by msfschaffner** [comment]
@matutem @jdonjdon @sha-ron we may want to look at this as part of the foundry cleanup.

**Comment by matutem** [comment]
We used to pass a timescale directive at compile time. What this does is to assign the given timescale to any modules without an explicit timescale, thus avoiding the viral timescale contagion from any random module that has a weird timescale. vcs and xcelium may require slightly different compilation flags for this.

**Comment by jdonjdon** [comment]
For the test bench / env, I vote for option1.  #11318 shows good example of how to do it.

**Comment by matutem** [comment]
@jdonjdon That means we would need to find all places that have timescale problems. The alternative I propose sets a default timescale for any module that has no explicit timescale. Notice this is different from alternative 2, and would avoid problems due to modules that set their own timescales.

This kind of problem is typically due to cell libraries, and it is better not to touch them. Well written cell libraries use a `resetall directive at the end or around each cell, which has the effect of cancelling any `timescale they set, but this bug shows we cannot expect all libraries to do that.

**Comment by jdonjdon** [comment]
@matutem I see your point. 
I thought we have that already in [here](https://cs.opensource.google/opentitan/opentitan/+/master:hw/dv/tools/dvsim/dsim.hjson;drc=a886c7eb73db63a9190d11e26de3edef37a35a80;l=20) or are you talking about another tool specific knob?

**Comment by matutem** [comment]
I am glad you asked. This is VCS specific, but here we go: the VCS documentation for -timescale says:

-timescale=time_unit/time_precision
Occasionally, some source files contain the ‘timescale
compiler directive and others do not. In this case, if you specify
the source files that do not contain the ‘timescale compiler
directive on the command line before you specify the ones that
do, this is an error condition and VCS halts compilation, by default.
This option enables you to specify the timescale for the source
files that do not contain this compiler directive and precede the
source files that do. Do not include spaces when specifying the
arguments to this option.

So you are right, this won't do, because order still matters. I think in a prior project we ended up adding `resetall directives to the library files, which was very unpleasant.

However, I found a few other useful flags for VCS:

-noinherit_timescale
This is a VCS or vlogan option that allows you to specify a global
timescale within the compilation unit of any source file that does
not have an explicit ‘timescale directive.

It seems that should do the job.

In terms of methodology, I think we could include a task to any tb.sv that checks the timescale is 1ns/1ps using $realtime at time 0, and kills the simulation if it fails. We can check that this task is effective with a simple experiment: set a timescale of 1ps/1ps in some commonly used prim, and expect the task to cause termination.

There are two more things to diagnose timescale issues:

- VCS will print the final timescale when finishing parsing in the build log.
- VCS provides a compile-time flag to diagnose timescale problems: `-diag timescale`.
